torch/fx/experimental/proxy_tensor.py

```diff
@@ -0,0 +1,207 @@
+    ) -> Any:
+        return forward(*args, **kwargs)
+
+    def _module_getattr(self, attr, attr_val, parameter_proxy_cache):
+        if isinstance(attr_val, torch.nn.Parameter):
+            for n, p in self.root.named_parameters():
+                if attr_val is p:
+                    if n not in parameter_proxy_cache:
+                        proxy = self.create_proxy('get_attr', n, (), {})
+                        parameter_proxy_cache[n] = ProxyTensor(attr_val, proxy)
+                    return parameter_proxy_cache[n]
+            return attr_val
+        return attr_val
+
+    # We need to do this so that parameters entering the `make_fx` context have
+    # a reference to them (and also have requires_grad set on them correctly
+    # I'm not actually sure if this is the right thing to do ...
+    def create_arg(self, a: Any):
+        if isinstance(a, torch.nn.Parameter):
+            for n, p in self.root.named_parameters():
+                if a is p:
+                    return self.create_node('get_attr', n, (), {})
+            qualname: Optional[str] = None
+
+            if not qualname:
+                i = 0
+                while True:
+                    qualname = f'_param_constant{i}'
+                    if not hasattr(self.root, qualname):
+                        break
+                    i += 1
+                setattr(self.root, qualname, a)
+
+            return self.create_node('get_attr', qualname, (), {})
+        return super().create_arg(a)
+
+
+def dispatch_trace(
+    root: Union[torch.nn.Module, Callable], concrete_args: Optional[Dict[str, Any]] = None
+) -> GraphModule:
+    tracer = PythonKeyTracer()
+    graph = tracer.trace(root, concrete_args)
+    name = root.__class__.__name__ if isinstance(root, torch.nn.Module) else root.__name__
+    return GraphModule(tracer.root, graph, name)
+
+
+def wrap_key(f, inps):
+    flat_inps, _ = pytree.tree_flatten(inps)
+
+    @functools.wraps(f)
+    def wrapped(*args):
+        flat_args, args_spec = pytree.tree_flatten(args)
+        assert(len(flat_args) == len(flat_inps))
+        for idx, arg in enumerate(flat_args):
+            if isinstance(flat_inps[idx], torch.Tensor):
+                flat_args[idx] = ProxyTensor(flat_inps[idx], arg)
+            else:
+                flat_args[idx] = flat_inps[idx]
+
+        tree_args = pytree.tree_unflatten(flat_args, args_spec)
+        out = f(*tree_args)
+        flat_outs, out_spec = pytree.tree_flatten(out)
+        for idx in range(len(flat_outs)):
+            if isinstance(flat_outs[idx], torch.Tensor) and isinstance(flat_outs[idx], ProxyTensor):
+                flat_outs[idx] = flat_outs[idx].proxy
+        return pytree.tree_unflatten(flat_outs, out_spec)
+
+    return wrapped
+
+
+def make_fx(f, decomposition_table={}):
+    @functools.wraps(f)
+    def wrapped(*args):
+        phs = pytree.tree_map(lambda x: fx.PH, args)
```

Comment:
this implies f cannot have any non-tensor arguments?

Reply:
yeah, `make_fx` does not support non-tensor arguments (or well, after pytree flattening).

Reply:
If we wanted to, we could simply bake those in if that's preferred.